autopush/waker.py

```diff
@@ -0,0 +1,101 @@
+    For instance, a vendor provided UDP wake extension for simplePush allows
+    servers to use a proprietary system that sends a UDP based wake message
+    to devices. This message bears no data or version information and only
+    wakes a device that may not be actively connected to the network.
+
+    """
+
+    def __init__(self, *args, **kwargs):
+        """ Setup """
+
+    def register(self, wake_info=None):
+        """ Endpoint Registration """
+
+    def set_active(self):
+        """Set period to time out and the destruction method"""
+
+    def check_idle(self):
+        """ Check if the connection has idled out """
+
+    def send_wake(self, wake_info=None):
+        """ Send a wake notification to the device """
+
+
+class UDPWake(IWake):
+    """ Proprietary UDP wake tool """
+
+    data = {}
+    default_period = 10 * 1000  # default to 10 seconds
+    endpoint = None     # Third party wake endpoint
+    timeout = 0
+    idle = 0
+
+    def _ms_time(self):
+        return int(time.time() * 1000)
+
+    def __init__(self, protocol=None, timeout=0, kill_func=None,
+                 host=None, cert=None, **kwargs):
+        self.idler = None
+        self.protocol = protocol
+        self.endpoint = host
+        self.cert = cert
+        if timeout > 0 and (protocol is None or kill_func is None):
+            raise ValueError("No protocol or kill function specified")
+        self.timeout = timeout
+        self.kill_func = kill_func
+        self.kill_args = kwargs
+
+    def register(self, info):
+        """Set the Wake info, usually done in 'hello'. This isn't really
+           useful for UDP, since the socket and the endpoint don't share
+           waker objects.
+        """
+        pass
+
+    def set_active(self):
+        """reset the idle timer"""
+        self.idle = self._ms_time()
+
+    def check_idle(self):
+        """Check to see if we need to kill the link"""
+        if self.idler is not None:
+            self.idler.cancel()
+        try:
+            if self.timeout > 0 and self.idle > 0:
+                if self._ms_time() - self.idle >= self.timeout * 1000:
+                    self.kill_func(**self.kill_args)
+                    return
+                self.idler = self.protocol.deferToLater(self.timeout,
+                                                        self.check_idle)
+        except (KeyError, AttributeError):
+            # More than likely, this isn't a UDP wake connection.
+            pass
+
+    def send_wake(self, wake_info=None):
+        """Send a wake request to the external endpoint."""
+        if wake_info is None:
+            wake_info = self.info
+        response = requests.post(self.endpoint,
+                                 data=urlencode(wake_info["data"]),
+                                 cert=self.cert)
+        if response.status_code < 200 or response.status_code >= 300:
+            raise WakeException()
```

Comment:
It seems this object does two things: maintains an idle timer on the connection, and sends a ping if the client is offline. Maybe it makes sense to break it up into two separate objects?

An idle timer could have other uses, too; at one time, I think we talked about disconnecting clients that don't say hello. Twisted's [`TimeoutMixin`](https://twistedmatrix.com/documents/current/api/twisted.protocols.policies.TimeoutMixin.html) might help here; it manages a timer object, and calls a ` timeoutConnection` method when the timer fires.

Second, instead of the endpoint, we could have `AutopushSettings` create and hold on to the `UDPWake` object (maybe in a `ap_settings.wakers` dict, if we wanted to get fancy). The `IWake` interface could then look like this:

* `sleep(self, proto)`: Called by `timeoutConnection`. This is basically your `kill_func`; it'd just call `proto.sendClose(4774, "UDP Idle")`.
* `wake(self, wake_info)`: Called by the endpoint when it wants to wake the client, after routing is complete.

That's just an idea, though. I'll ponder it some more.